DEVELOPMENT_CONTRIBUTION.md

```diff
@@ -0,0 +1,159 @@
+git config user.name "John Doe"
+git config user.email johndoe@example.com
+
+## Setup the upstream remote
+git remote add upstream https://github.com/vmware/clarity.git
+
+## Check out the upstream a topic branch for your changes
+git fetch
+git checkout -b topic/feature-name upstream/topic/feature-name
+```
+
+### Commits
+
+Open-source for web UI still feels like the Wild West, so we're trying to take a page from older open-source projects, even if... _Javascript_ 🙄.  In particular, if your contribution is large, split it into smaller commits that are logically self-contained. You can then submit them as separate pull requests that we will review one by one and merge progressively into the topic branch. As a rule of thumb, try to keep each pull request under a couple hundred lines of code, _unit tests included_. We realize this isn't always easy, and sometimes not possible at all, so feel free to ask how to split your contribution in the GitHub issue. In general, it's a good idea to start coding the services first and test them in isolation, then move to the components.
+
+For your commit message, please use the following format:
+```
+[<type>] <title>
+<BLANK LINE>
+<detailed commit message>
+<BLANK LINE>
+<reference to closing an issue>
+<BLANK LINE>
+Signed-off-by: Your Name <your.email@example.com>
+```
+
+Type must be one of the following:
+
+- `UX`: Research and design principles. These are typically documentation commits.
+- `ICON`: Implementation of Clarity's Custom Element icons.
+- `UI`: Implementation of Clarity's static styles, its general look-and-feel.
+- `NG`: Implementation of Clarity's Angular components.
+- `WEB`: Website and documentation updates.
+- `CORE`: Distribution, build script and tooling.
+
+These documents provide guidance creating a well-crafted commit message:
+
+- [How to Write a Git Commit Message](http://chris.beams.io/posts/git-commit/)
+- [Closing Issues Via Commit Messages](https://help.github.com/articles/closing-issues-via-commit-messages/)
+
+
+### Submitting pull requests 
+
+As you implement your contribution, make sure all work stays on your local topic branch. When an isolated part of the feature is complete with unit tests, make sure to submit your pull request **against the topic branch** on the main Clarity repository instead of `master`. This will allow us to accept and merge partial changes that shouldn't make it into a production release of Clarity yet. We expect every pull request to come with exhaustive unit tests for the submitted code.
+
+**Do not, at any point, rebase your local topic branch on newer versions of `master` while your work is still in progress!** This will create issues both for you, the reviewers, and maybe even other developers who might submit additional commits to your topic branch if you requested some help.
+
+To make sure your pull request will pass our automated testing, before submitting you should:
+
+- Make sure `npm test` passes for each of them.
```

Comment:
We should update `npm test` to include formatting checks so this is easier.

Reply:
Actually, we didn't include formatting/linting in `npm test` on purpose. We used to have it, and your tests would keep failing as you were debugging them because you didn't leave a new blank line somewhere, or because you used the wrong quotes. It was a nightmare, we want to keep these processes separate to not hinder test writing/debugging.